indico/modules/events/registration/clone.py

```diff
@@ -22,48 +22,88 @@
-            items_attrs = get_simple_column_attrs(RegistrationFormItem) | {'parent_id'}
+                                                                                 for attr in items_attrs})
+            old_section_items = RegistrationFormItem.find(parent=old_section)
             for old_item in old_section_items:
                 new_item = RegistrationFormItem(parent=new_section, registration_form=new_form,
                                                 **{attr: getattr(old_item, attr) for attr in items_attrs})
                 if new_item.is_field:
-                    field_data = RegistrationFormFieldData(field=new_item,
-                                                           versioned_data=getattr(old_item, 'versioned_data'))
-                    new_item.current_data = field_data
+                    if self._clone_with_registrations:
+                        self._clone_all_field_versions(old_item, new_item)
+                    else:
+                        field_data = RegistrationFormFieldData(field=new_item,
+                                                               versioned_data=old_item.versioned_data)
+                        new_item.current_data = field_data
                 new_section.children.append(new_item)
             db.session.add(new_section)
             db.session.flush()
+
+    def _clone_all_field_versions(self, old_item, new_item):
+        for old_version in old_item.data_versions:
+            new_version = RegistrationFormFieldData(versioned_data=old_version.versioned_data, field=new_item)
+            if old_version.id == old_item.current_data_id:
+                new_item.current_data = new_version
+            new_item.data_versions.append(new_version)
+            self._field_data_dict[old_version] = new_version
+
+    def _clone_registrations(self, old_form, new_form, new_event):
+        registration_attrs = get_simple_column_attrs(Registration) - {'uuid', 'ticket_uuid'}
+        for old_registration in old_form.registrations:
+            new_registration = Registration(event_id=new_event.id, registration_form=new_form,
+                                            user=old_registration.user, **{attr: getattr(old_registration, attr)
+                                                                           for attr in registration_attrs})
+            reg_data_attrs = get_simple_column_attrs(RegistrationData)
+            for old_registration_data in old_registration.data:
+                new_registration_data = RegistrationData(registration=new_registration,
+                                                         **{attr: getattr(old_registration_data, attr)
+                                                            for attr in reg_data_attrs})
+                new_registration_data.field_data = self._field_data_dict[old_registration_data.field_data]
+                new_registration.data.append(new_registration_data)
+            new_form.registrations.append(new_registration)
```

Comment:
You can remove either this line or the `registration_form=new_form` when creating the registration (those two are relationship+backref and synced automatically)